---
title: "TCGA analysis"
author: "Paulyna Magana"
output:
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: true
    fig_width: 6
    fig_height: 8
    fig_caption: true
header-includes:
- \usepackage{gb4e}
- \noautomath


knit: (function(input_file, encoding) { output_dir = "./";
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), output_dir, 'TCGA_analysis'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

## TCGA Analysis

```{r}
library("TCGAbiolinks")
library("limma")
library("edgeR")
library("caret")
library("SummarizedExperiment")
library("gplots")
library("survival")
library("survminer")
library("RColorBrewer")
library("genefilter")
```

Query the TCGA database through R with the function GDCquery.

Check all the available projects at TCGA with the command bellow.


```{r}
GDCprojects = getGDCprojects()

head(GDCprojects[c("project_id", "name")] )
```


```{r}
dplyr::filter(GDCprojects, grepl('Lung', name))
```

# For TCGA-LUSC, get details on all the data deposited


```{r}
TCGAbiolinks:::getProjectSummary("TCGA-LUAD")
```

Of note, not all patients were measured for all data types. Also, some data types have more files than samples. This is the case when more experiments were performed per patient, i.e. transcriptome profiling was performed both in mRNA and miRNA, or that data have been analysed by distinct computational strategies.

Let us start by querying all RNA-seq data from LUAD project.

When using GDCquery we always need to specify the id of the project, i.e. “TCGA_LUSC”, and the data category we are interested in, i.e. “Transcriptome Profiling”. Here, we will focus on a particular type of data summarization for mRNA-seq data (workflow.type), which is based on raw counts estimated with HTSeq.

Note that performing this query will take a few of minutes


```{r}
query_TCGA = GDCquery(
  project = "TCGA-LUAD",
  data.category = "Transcriptome Profiling", # parameter enforced by GDCquery
  experimental.strategy = "RNA-Seq") # some cases there might be more than 1 data type = gene expression, splice junstion quantification
```

To visualize the query results in a more readable way, we can use the command getResults.


```{r}
lusc_res = getResults(query_TCGA)

colnames(lusc_res)
```


```{r}
head(lusc_res)
```


```{r}
summary(factor(lusc_res$sample_type))
```

In some cases, there might be more than one data_type(as column in table).
I'm querying the Gene expression Quantification but there is also Splice Junction Quantification

```{r}
query_TCGA_raw = GDCquery(
  project = "TCGA-LUAD",
  data.category = "Transcriptome Profiling", # parameter enforced by GDCquery
  access= "open",
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  data.type = "Gene Expression Quantification") #double check
```



```{r}
GDCdownload(query_TCGA_raw, method = "api")
```


It might take some time, and be downloaded in chunks to a folder called GDCdata

```{r}
luad.count  <- GDCprepare(query = query_TCGA_raw, save = TRUE, save.filename = "TCGA_LUAD.rda", summarizedExperiment = TRUE)

```



```{r}
head(query_TCGA_raw)

```


Lets load the actual RNASeq data into R. Remember that the output directory set must be the same to where you downloaded the data.


















